### PR TITLE
Improve experience in nested container environments

### PR DESCRIPTION
**Pull request by baez90**
# What does this PR do?

## Container env detection

Currently testcontainers-go detects only nested Docker environments (no check for `/run.container` for Podman) and from my understanding it only checks for the default gateway of the default network to use this for port mappings.

This PR refactors the current behavior to also consider Podman environments and to detect its own container ID from the `/etc/hostname` mount (both Docker and Podman are doing this somehow similar with a tmpfs mount from a directory that has the container ID in the mount root path.

## Bind mount remapping

The library now detects if it is running in a container (with a local Docker daemon i.e. with a Docker socket, not a TCP/HTTP connection) and if a developer attempts to use a bind mount in this setting it will try to map the source path to a host path.

Considering the following example:

- CI provider clones the repo to `/home/ci/work/1234/` on the __host__
- spawns a new container like so: `docker run --rm -i -v /home/ci/work/1234:/work -v /var/run/docker.sock:/var/run/docker.sock:ro -w /work ...`
- __Scenario A__: 
    - test code attempts to bind mount `/work/testdata/my-config.toml` to a test container
    - `tc-go` detects it is running in container environment and that the path `/work/` is actually part of a bind mount
    - `tc-go` "re-maps" the mount source `/work/testdata/my-config.toml` to `/home/ci/work/1234/testdata/my-config.toml`
- __Scenario B__: 
    - test code attempts to bind mount `/etc/test/my-config.toml` to a test container
    - `tc-go` detects it is running in container environment and that the path `/etc/test/my-config.toml` might not be available if mounted to another container and returns an error

There are some corner cases that are currently not yet taken care of like:
- a user might intentionally mount a path that is not part of a bind mount
- there's a TCP-to-socket proxy in place i.e. remote daemon detection doesn't work
- ...

### CODE REVIEW

**Review by mdelapenya — COMMENTED** [review]
I left a few minor comments before testing the changes locally.

Could you please resolve the go.mod conflicts too?

File: docker.go [hunk 1]
```diff
@@ -1377,3 +1344,156 @@ func (p *DockerProvider) getDefaultNetwork(ctx context.Context, cli *client.Clie
...
+}
+
+func (p *DockerProvider) initContainerEnvInformation(ctx context.Context) error {
+	var initErr error
+
+	p.initContainerEnvOnce.Do(func() {
+		remote, err := p.isConnectedToRemoteDaemon()
+		if err != nil {
+			initErr = err
+			return
```
[the remark this answers is not in the document]
Comment (baez90) [reply docker.go:1438]: :smile: yeah that is not really a beauty but it's a limitation of `sync.Once`.
The `initErr` is declared outside of the `sync.Once` closure and is returned at the very end.
Every error occurring within the `sync.Once` closure is assigned to the `initErr` to allow the outer function returning the error in case there was one.
The `return` in the next line does not return the error but only returns from the closure.

You know what I mean?

File: docker.go [hunk 2]
```diff
@@ -1377,3 +1344,156 @@ func (p *DockerProvider) getDefaultNetwork(ctx context.Context, cli *client.Clie
...
+			containerMount.TmpfsOptions = typedMounter.GetTmpfsOptions()
+		}
+
+		mounts = append(mounts, containerMount)
+	}
+
+	return mounts, nil
+}
+
+func (p *DockerProvider) initContainerEnvInformation(ctx context.Context) error {
```
[the remark this answers is not in the document]
Comment (baez90) [reply docker.go:1431]: will do!

File: wait/testdata/go.mod [hunk 3]
```diff
@@ -1,3 +1,3 @@
 module httptest
 
-go 1.15
+go 1.19
```
[the remark this answers is not in the document]
Comment (baez90) [reply wait/testdata/go.mod:3]: Can do, honestly I had some weird issues, was trying some things and forgot to revert it and after noticing in the PR overview I figured it wouldn't really matter as it is only for testing but can do :blush:

File: docker.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mdelapenya) [reply docker.go:1438]: Ah I got it, Probably because I was using GH for the review I did not understand each scope properly. Thanks for the clarification!

[the remark this answers is not in the document]
Comment (mdelapenya) [reply docker.go:1438]:
```golang
p.initContainerEnvOnce.Do(func() {
```

It's a function 🤦  Thanks again for your patience

**Comment by baez90** [comment]
I'll resolve the conflicts in the evening.
I wasn't sure how to 'fix' the coverage issue as I don't have much experience with codecov and it would be necessary to somehow 'merge' the coverage results from the nested test run and from the non-nested one. Do you happen to have an idea how to do this? Otherwise I'll give it a try also this evening :blush: 

Congratulations to the new position! :tada:  I'm excited to hear that the community got such a great supporter with you! Wish you all the best with this new challenge :blush:

File: docker.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (baez90) [reply docker.go:1438]: Yeah, the API of `sync.Once` is not that great :smile: 
I would love to have something like `sync.Once{}.DoSuccessfully(func() error{ ... }))` to be able to repeat it if it erros or so but yeah....maybe in the future :smile: but I totally get the confusion especially when using GitHub for reviewing

**Comment by mdelapenya** [comment]
> I'll resolve the conflicts in the evening. I wasn't sure how to 'fix' the coverage issue as I don't have much experience with codecov and it would be necessary to somehow 'merge' the coverage results from the nested test run and from the non-nested one. Do you happen to have an idea how to do this? Otherwise I'll give it a try also this evening 😊

Yeah, many times codecov complains about a line not being touched, even on dependabot PRs, which makes me think if it's adding more noise than value. So I'd delegate a final approval on the coverage to a human

File: docker.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (baez90) [reply docker.go:1436]: pls let me know if this sufficient?

[the remark this answers is not in the document]
Comment (mdelapenya) [reply docker.go:1436]: Yes, thank you!!

**Comment by baez90** [comment]
Don't know how to fix the patch coverage issue but at least to overall coverage is fixed 🙂
Let me know if there's something missing otherwise I'd squash the commits to make this ready to be merged?

**Review by mdelapenya — DISMISSED** [review]
I added a few ultra-minor comments, but other than that LGTM

I noticed there are changes that could have been sent in a separate PR (replacement of ioutil/io, go version in module files, indents, filepath for the win, usage of constants in tests, better assertions, better usage of t.Testing functions -helpers,environment vars-, increased timeouts) but I'm pretty sure they we re discovered and needed while working in the nested checks.

For further PRs I'd appreciate if we could send those small chunks separately, so that the review of the original feature (nested container environments) could be done faster 🙏 

Thanks again for your contribution to make tc-go great!

File: container.go [hunk 4]
```diff
@@ -264,3 +265,12 @@ func (c *ContainerRequest) validateMounts() error {
 	}
 	return nil
 }
+
+func RunningInContainer() bool {
+	return fileExists("/.dockerenv") || fileExists("/run/.containerenv")
```
[the remark this answers is not in the document]
Comment (baez90) [reply container.go:270]: If I remember correctly it is not only used in testing but also to determine if the container environment information should be fetched.

I was wondering as well if it makes sense to expose this but I thought it could be useful and it's not a huge 'burden' to maintain :smile:

Regarding other indicators: I'm not sure how Docker is handling the same with Windows containers so there might be another check in the future to cover these as well (although supporting Windows containers will be fun anyway if this is something that should be done in the future) but apart from that I hope `/run/.containerenv` is not only used by Podman but by all other container providers apart from Docker

File: container_test.go [hunk 5]
```diff
@@ -335,6 +334,7 @@ func TestShouldStartContainersInParallel(t *testing.T) {
 }
 
 func createTestContainer(t *testing.T, ctx context.Context) int {
+	t.Helper()
```
[the remark this answers is not in the document]
Comment (baez90) [reply container_test.go:337]: I had a look but haven't seen any. Of course this only makes sense if you pass a `testing.TB` implementation and most other utilities don't get one hence I didn't touch them for now

**Comment by baez90** [comment]
I'll have a look at your nits probably tomorrow, thanks for your feedback!

Sorry about "mixing" up the PR, I tried to separate the changes in multiple commits to improve the review flow e.g. all the `ioutil` changes are in a separate commit and so but I'll try to narrow the scope in the future, sorry again :sweat_smile:

**Comment by mdelapenya** [comment]
> Sorry about "mixing" up the PR, I tried to separate the changes in multiple commits to improve the review flow e.g. all the `ioutil` changes are in a separate commit and so but I'll try to narrow the scope in the future, sorry again 😅

No worries at all, I totally understand it and I can imagine myself adding commits on top as you did while finding "things". So not a problem :)

I mentioned because we are tying to have a better internal review process, and things like that would improve our performance 😅

File: container.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (kiview) [reply container.go:270]: Be careful about `.dockerenv`. To my last knowledge, this is an undocumented implementation detail of Docker and not necessarily a stable indicator. There is some discussion around this on SO, which sadly does not provide a better alternative: https://superuser.com/questions/1021834/what-are-dockerenv-and-dockerinit

**Comment by kiview** [comment]
I am joining this PR as a `tc-java` maintainer and I might the bigger picture, just wanted to point out something I noticed in the description of the PR:

> I'm also planning to rework the mount behavior a bit to auto-replace bind mounts source paths to make them host-path relative if possible or raise an error if the mount won't work due to nested environment constraints

There are other reasons why the mount won't work, most notably a remote Docker daemon. With making them host-path relative are you specifically talking about the Docker-Socket-Mounting scenario? I believe solving this can be a bigger headache than clearly not supporting it and using copy-to-container mechanisms instead.

**Comment by mdelapenya** [comment]
> using copy-to-container mechanisms instead.

Thanks @kiview, I was planing to articulate exactly this into a separate issue, maybe it makes sense to add it here as a reference.

**Comment by baez90** [comment]
Thanks @kiview for the feedback.
I honestly forgot to update the PR description but you're right, this is only about re-mapping bind mounts if the current process is running in a container and the docker daemon is __not__ a remote daemon.

Of course this could as well left out but I'm hitting this scenario a lot in the last time and always building custom images only to work around this issue in certain circumstances means a significant loss of performance in comparison to just re-map the source if possible - especially because you've to build the image no matter if running locally or in the CI.
Also I think it improves the user experience if an error is returned in case testcontainers-go detected that a re-mapping is not possible instead of just returning an error from the Docker daemon telling the user a directory doesn't exist although it very well does if the test is executed locally but not in the CI environment just because e.g. Microsoft decided to mount the working directory in another path in the container than on the host.

I don't intent to solve this issue altogether (like also taking care of volume mounts....) but in this special scenario it...doesn't seem like such a headache to be honest? :sweat_smile:

**Comment by kiview** [comment]
@baez90 Thanks for the explanation, I was initially missing the meaning of `remap` in this context. I agree that this likely helps with the DX (developer experience) if the test fails early in environments where this is not supported. I am simply a bit cautious regarding how robust the detection is (because I have seen so weird Docker environments from users in `tc-java`). Probably I need to look more in the `tc-go` implementations around this, to also learn if there are things we can apply to `tc-java` as well.

**Comment by baez90** [comment]
:smile: yeah I mean there are always things that can be improved and some corner cases that are hard to circumvent e.g. if a user has a...socket-to-TCP proxy to - for whatever reason - not use the TCP connection directly it'll be nearly impossible to detect this but for such scenarios it's still possible to fallback to copy-to-container or introduce an option to explicitly disable the mapping later. Also there are of course still scenarios that could be covered to further improve the DX but this PR is already blown up too much :smile: so probably something for later

File: container.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (baez90) [reply container.go:270]: I am ware, although Podman explicitly adopted this with their `/run/.containerenv` flag.
Surely this is not the greatest solution but a fairly easy one and for instance also used in [`tc-dotnet`](https://github.com/testcontainers/testcontainers-dotnet/blob/ff76e0fb4ff00bc81b657fda982ab8c1608f6fd3/src/Testcontainers/Clients/TestcontainersClient.cs).
Furthermore the SO link you attached mentions a moby repository where they added the [exact same behavior](https://github.com/moby/libnetwork/blob/b3507428be5b458cb0e2b4086b13531fb0706e46/drivers/bridge/setup_bridgenetfiltering.go#L160) so it seems this flag is considered stable enough by the Docker guys :smile: 

An alternative approach could be to not use these flags at all but to check if `/etc/hostname` is mounted (by checking the `/proc/self/mountinfo`) and if so try to get the container ID from the mount root.

Both solutions are not perfect, I'm open for suggestions

**Comment by baez90** [comment]
What's the current state of this? Is it missing something?

**Comment by mdelapenya** [comment]
@baez90 after moving compose to a separate module, this PR got into conflicts. Are you still interested on it? Otherwise we could close it and eventually submit a new one.

Thanks!

**Comment by mdelapenya** [comment]
Hey @baez90 I'm tempted to close this PR because of

>I'm actually wondering if it would not make sense to re-create this PR instead of trying to rebase it.
Also I currently do not have the use case for it anymore.

Just to let you know, there is a WIP PR for improving podman support here: https://github.com/testcontainers/testcontainers-go/pull/1251

**Comment by baez90** [comment]
Yeah, I won't be able to progress on that in the close future I'm afraid, thanks!
